### ISSUE TITLE
[rllib]RNN support for SAC

### ISSUE THREAD

**Issue by ddworak94** [ray-project/ray#16014]
Me and my team was lacking an option to train SAC algorithm with support for RNNs, so i've implemented a RNNSAC trainer with a relevant RNNSAC policy and model that more or less bases on original SAC agent implementation and I also took inspiration from R2D2 implementation for RNNs support. I've added support for hidden states in RNN based policy and q models, burn_in and zero_init_states, as well as use_prev_reward/action TrajectoryView api. Tested it on repeat_after_me and stateless_cartpole env. 

Here is my forked repo: https://github.com/ddworak94/ray
Main contribution is implementation of rllib/agents/sac/rnnsac.py , rnnsac_torch_policy.py and rnnsac_torch_model.py.
I've also added registry record for this model and lastly a thing i would consider a bug:
in rllib/policy/rnn_sequencing.py:388
`is_last_episode_ids[0] is not True` yields True even if first element of the list is True due to np._True type i think. 
Casting it to bool() fix this and allow init_states to be correctly passed to the sequance.

I was wondering if I could get an opinion from someone in dev team whether this feature looks alright and could be integrated into the master via pull request or some further adjustments needs to be done?

**Comment by mvindiola1** [comment]
@ddworak94 Nice work. I cannot actually approve anything that is @sven1977. I am sure he will have some feedback. As I was looking through the code three things jumped out at me that I thought I would comment on.

1. I did not see the tf implementation. Did I miss it or did you just add a torch version?
2. Nice catch on this one `if bool(is_last_episode_ids[0]) is not True`! I was looking at that code a couple weeks ago and missed it so kudos to you. I think the fix can be simpler though; wouldn't either of these work fine? `if not is_last_episode_ids[0]` or `if is_last_episode_ids[0] == False:`
3. I noticed in the loss function that you are both multiplying by a boolean mask seq_mask and reduce_mean_valid. You do not need to do both of these. The  reduce_mean_valid function will only select the valid rows so multiplying by the mask is redundant.

Again, thanks for the contribution.

**Comment by ddworak94** [comment]
@mvindiola1 Thank you for your review and comments. Answering your questions:

1. Yes, I have only implemented torch version so far. I guess if everything else is correct, implementing tf model and policy shouldn't be much of a problem.
2. Sure `if not is_last_episode_ids[0]` looks way better then my explicit casing, I just wanted to highlight what causes the problem.
3. Following your suggestion i've changed the code to use reduce_mean_valid only, except i need to 'manually' mask also things saved in the stats like td_error (right?)

Also i have question regarding model config keys: "batch_mode" & "prioritized_replay".
Those 2 options are frozen for R2D2(rnn) to "complete_episodes" and "False", but it seems like i could use "truncate_episodes" and prioritized replay buffer in RNNSAC. The question is does it really work underneath or does it just passes the checks?

**Comment by ddworak94** [comment]
Hello,

Its been 2 weeks since original post and i was wondering if i can get an update?
Our goal is to include such agent (SAC with RNN support) to official repo in order to maintain support for future changes.
Do you think it is okey to be merge or some additional work needs to be done (I can help with that also).

**Comment by sven1977** [comment]
Hey @ddworak94 , sorry for the delay. This looks super cool in general :) Could you PR this?
Don't worry about tf missing, we can add this later.

**Comment by mch5048** [comment]
@ddworak94 Nice work!
